File: frontend/src/lib/types/arm.ts
import type { TranscoderJob } from './transcoder';

export interface Track {
	track_id: number;
	job_id: number;
	track_number: string | null;
	length: number | null;
	aspect_ratio: string | null;
	fps: number | null;
	enabled: boolean | null;
	basename: string | null;
	filename: string | null;
	orig_filename: string | null;
	new_filename: string | null;
	ripped: boolean | null;
	status: string | null;
	error: string | null;
	source: string | null;
	// Per-track title metadata (null = inherits from job)
	title: string | null;
	year: string | null;
	imdb_id: string | null;
	poster_url: string | null;
	video_type: string | null;
	// TVDB episode matching
	episode_number: string | null;
	episode_name: string | null;
	// User-specified output filename (overrides pattern rendering)
	custom_filename: string | null;
}

export interface Job {
	job_id: number;
	arm_version: string | null;
	crc_id: string | null;
	logfile: string | null;
	start_time: string | null;
	stop_time: string | null;
	job_length: string | null;
	status: string | null;
	stage: string | null;
	no_of_titles: number | null;
	title: string | null;
	title_auto: string | null;
	title_manual: string | null;
	year: string | null;
	year_auto: string | null;
	year_manual: string | null;
	video_type: string | null;
	video_type_auto: string | null;
	video_type_manual: string | null;
	imdb_id: string | null;
	imdb_id_auto: string | null;
	imdb_id_manual: string | null;
	poster_url: string | null;
	poster_url_auto: string | null;
	poster_url_manual: string | null;
	devpath: string | null;
	mountpoint: string | null;
	hasnicetitle: boolean | null;
	errors: string | null;
	disctype: string | null;
	label: string | null;
	path: string | null;
	raw_path: string | null;
	transcode_path: string | null;
	artist: string | null;
	artist_auto: string | null;
	artist_manual: string | null;
	album: string | null;
	album_auto: string | null;
	album_manual: string | null;
	season: string | null;
	season_auto: string | null;
	season_manual: string | null;
	episode: string | null;
	episode_auto: string | null;
	episode_manual: string | null;
	transcode_overrides: Record<string, unknown> | null;
	multi_title: boolean | null;
	title_pattern_override: string | null;
	folder_pattern_override: string | null;
	disc_number: number | null;
	disc_total: number | null;
	ejected: boolean | null;
	pid: number | null;
	manual_pause: boolean | null;
	source_type?: string;
	source_path?: string;
	wait_start_time: string | null;
	tracks_total: number | null;
	tracks_ripped: number | null;
	tvdb_id: number | null;
}

export interface JobDetail extends Job {
	tracks: Track[];
	config: Record<string, string | null> | null;
}

export interface JobListResponse {
	jobs: Job[];
	total: number;
	page: number;
	per_page: number;
	pages: number;
}

export interface SystemInfo {
	id: number;
	name: string | null;
	cpu: string | null;
	description: string | null;
	mem_total: number | null;
}

export interface HardwareInfo {
	cpu: string | null;
	memory_total_gb: number | null;
}

export interface Drive {
	drive_id: number;
	name: string | null;
	mount: string | null;
	job_id_current: number | null;
	job_id_previous: number | null;
	description: string | null;
	drive_mode: string | null;
	maker: string | null;
	model: string | null;
	serial: string | null;
	connection: string | null;
	read_cd: boolean | null;
	read_dvd: boolean | null;
	read_bd: boolean | null;
	firmware: string | null;
	location: string | null;
	stale: boolean | null;
	mdisc: number | null;
	serial_id: string | null;
	uhd_capable: boolean | null;
	rip_speed: number | null;
	prescan_cache_mb: number | null;
	prescan_timeout: number | null;
	prescan_retries: number | null;
	disc_enum_timeout: number | null;
	current_job: Job | null;
}

export interface Notification {
	id: number;
	title: string | null;
	message: string | null;
	trigger_time: string | null;
	seen: boolean;
	cleared: boolean;
}

export interface MemoryInfo {
	total_gb: number;
	used_gb: number;
	free_gb: number;
	percent: number;
}

export interface StoragePath {
	name: string;
	path: string;
	total_gb: number;
	used_gb: number;
	free_gb: number;
	percent: number;
}

export interface GpuSnapshot {
	vendor: string;
	utilization_percent: number | null;
	memory_used_mb: number | null;
	memory_total_mb: number | null;
	temperature_c: number | null;
	encoder_percent: number | null;
	power_draw_w: number | null;
	power_limit_w: number | null;
	clock_core_mhz: number | null;
	clock_memory_mhz: number | null;
}

export interface SystemStats {
	cpu_percent: number;
	cpu_temp: number;
	memory: MemoryInfo | null;
	storage: StoragePath[];
	gpu: GpuSnapshot | null;
}

export interface DashboardData {
	db_available: boolean;
	arm_online: boolean;
	active_jobs: Job[];
	system_info: HardwareInfo | null;
	drives_online: number;
	drive_names: Record<string, string>;
	notification_count: number;
	ripping_enabled: boolean;
	makemkv_key_valid: boolean | null;
	makemkv_key_checked_at: string | null;
	transcoder_online: boolean;
	transcoder_stats: Record<string, unknown> | null;
	transcoder_system_stats: SystemStats | null;
	active_transcodes: TranscoderJob[];
	system_stats: SystemStats | null;
	transcoder_info: HardwareInfo | null;
}

export interface LogFile {
	filename: string;
	size: number;
	modified: string;
}

export interface LogContent {
	filename: string;
	content: string;
	lines: number;
}

export interface LogEntry {
	timestamp: string;
	level: string;
	logger: string;
	event: string;
	job_id: number | null;
	label: string | null;
	raw: string;
}

export interface StructuredLogContent {
	filename: string;
	entries: LogEntry[];
	lines: number;
}

export interface SearchResult {
	title: string;
	year: string;
	imdb_id: string | null;
	media_type: string;
	poster_url: string | null;
}

export interface MediaDetail extends SearchResult {
	plot: string | null;
	background_url: string | null;
}

export interface MusicSearchResult {
	title: string;
	artist: string;
	year: string;
	release_id: string;
	media_type: string;
	poster_url: string | null;
	track_count: number | null;
	country: string | null;
	release_type: string | null;
	format: string | null;
	label: string | null;
}

export interface MusicDetail extends MusicSearchResult {
	catalog_number: string | null;
	barcode: string | null;
	status: string | null;
	disc_count: number | null;
	tracks: { number: string; title: string; length_ms: number | null; disc_number: number | null }[];
}

export interface TitleUpdate {
	title?: string;
	year?: string;
	video_type?: string;
	imdb_id?: string;
	poster_url?: string;
	path?: string;
	disctype?: string;
	label?: string;
	artist?: string;
	album?: string;
	season?: string;
	episode?: string;
}

export interface TrackTitleUpdate {
	title?: string;
	year?: string;
	video_type?: string;
	imdb_id?: string;
	poster_url?: string;
}

export interface JobConfigUpdate {
	RIPMETHOD?: 'mkv' | 'backup';
	DISCTYPE?: 'dvd' | 'bluray' | 'bluray4k' | 'music' | 'data';
	MAINFEATURE?: boolean;
	MINLENGTH?: number;
	MAXLENGTH?: number;
	AUDIO_FORMAT?: string;
	SKIP_TRANSCODE?: boolean;
}

export interface TranscoderConfig {
	config: Record<string, unknown>;
	updatable_keys: string[];
	paths?: { raw_path: string; completed_path: string; work_path: string };
	valid_video_encoders?: string[];
	valid_audio_encoders?: string[];
	valid_subtitle_modes?: string[];
	valid_log_levels?: string[];
	valid_handbrake_presets?: string[];
	valid_preset_files?: string[];
	presets_by_file?: Record<string, string[]>;
}

export interface TranscoderAuthStatus {
	require_api_auth: boolean;
	webhook_secret_configured: boolean;
}

export interface FolderScanResult {
	disc_type: string;
	label: string;
	title_suggestion: string;
	year_suggestion: string | null;
	folder_size_bytes: number;
	stream_count: number;
	disc_number?: number | null;
	disc_total?: number | null;
	season?: number | null;
}

export interface FolderCreateRequest {
	source_path: string;
	title: string;
	year?: string | null;
	video_type: string;
	disctype: string;
	imdb_id?: string | null;
	poster_url?: string | null;
	multi_title?: boolean;
	season?: number | null;
	disc_number?: number | null;
	disc_total?: number | null;
}

export interface FolderCreateResponse {
	job_id: number;
	status: string;
	source_type: string;
	source_path: string;
}

export interface SettingsData {
	arm_config: Record<string, string | null> | null;
	arm_metadata: Record<string, string> | null;
	arm_handbrake_presets: string[] | null;
	naming_variables: Record<string, string> | null;
	transcoder_config: TranscoderConfig | null;
	transcoder_gpu_support: Record<string, boolean> | null;
	transcoder_auth_status: TranscoderAuthStatus | null;
	/** @deprecated Use transcoder_gpu_support instead */
	gpu_support: Record<string, boolean> | null;
}

File: backend/models/arm.py
"""Read-only SQLAlchemy models mirroring the ARM database schema."""

from sqlalchemy import (
    Column,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    Boolean,
)
from sqlalchemy.orm import DeclarativeBase, relationship


class Base(DeclarativeBase):
    pass


class Job(Base):
    __tablename__ = "job"

    job_id = Column(Integer, primary_key=True)
    arm_version = Column(String(20))
    crc_id = Column(String(63))
    logfile = Column(String(256))
    start_time = Column(DateTime)
    stop_time = Column(DateTime)
    job_length = Column(String(20))
    status = Column(String(32))
    stage = Column(String(64))
    no_of_titles = Column(Integer)
    title = Column(String(256))
    title_auto = Column(String(256))
    title_manual = Column(String(256))
    year = Column(String(4))
    year_auto = Column(String(4))
    year_manual = Column(String(4))
    video_type = Column(String(20))
    video_type_auto = Column(String(20))
    video_type_manual = Column(String(20))
    imdb_id = Column(String(15))
    imdb_id_auto = Column(String(15))
    imdb_id_manual = Column(String(15))
    poster_url = Column(String(256))
    poster_url_auto = Column(String(256))
    poster_url_manual = Column(String(256))
    devpath = Column(String(15))
    mountpoint = Column(String(20))
    hasnicetitle = Column(Boolean)
    errors = Column(Text)
    disctype = Column(String(20))
    label = Column(String(256))
    path = Column(String(256))
    raw_path = Column(String(256))
    transcode_path = Column(String(256))
    ejected = Column(Boolean)
    updated = Column(Boolean)
    pid = Column(Integer)
    pid_hash = Column(Integer)

    tracks = relationship("Track", back_populates="job", lazy="selectin")
    config = relationship("Config", back_populates="job", uselist=False, lazy="selectin")


class Track(Base):
    __tablename__ = "track"

    track_id = Column(Integer, primary_key=True)
    job_id = Column(Integer, ForeignKey("job.job_id"))
    track_number = Column(String(4))
    length = Column(Integer)
    aspect_ratio = Column(String(20))
    fps = Column(Float)
    main_feature = Column(Boolean)
    basename = Column(String(256))
    filename = Column(String(256))
    orig_filename = Column(String(256))
    new_filename = Column(String(256))
    ripped = Column(Boolean)
    status = Column(String(32))
    error = Column(Text)
    source = Column(String(256))

    job = relationship("Job", back_populates="tracks")


class Config(Base):
    __tablename__ = "config"

    CONFIG_ID = Column(Integer, primary_key=True)
    job_id = Column(Integer, ForeignKey("job.job_id"))
    ARM_CHECK_UDF = Column(String(10))
    GET_VIDEO_TITLE = Column(String(10))
    SKIP_TRANSCODE = Column(String(10))
    VIDEOTYPE = Column(String(20))
    MINLENGTH = Column(String(10))
    MAXLENGTH = Column(String(10))
    MANUAL_WAIT = Column(String(10))
    MANUAL_WAIT_TIME = Column(String(10))
    RAW_PATH = Column(String(256))
    TRANSCODE_PATH = Column(String(256))
    COMPLETED_PATH = Column(String(256))
    EXTRAS_SUB = Column(String(256))
    INSTALLPATH = Column(String(256))
    LOGPATH = Column(String(256))
    LOGLEVEL = Column(String(20))
    LOGLIFE = Column(String(10))
    DBFILE = Column(String(256))
    WEBSERVER_IP = Column(String(20))
    WEBSERVER_PORT = Column(String(10))
    SET_MEDIA_PERMISSIONS = Column(String(10))
    CHMOD_VALUE = Column(String(10))
    SET_MEDIA_OWNER = Column(String(10))
    CHOWN_USER = Column(String(64))
    CHOWN_GROUP = Column(String(64))
    RIPMETHOD = Column(String(20))
    MKV_ARGS = Column(String(256))
    DELRAWFILES = Column(String(10))
    HASHEDKEYS = Column(String(256))
    HB_PRESET_DVD = Column(String(256))
    HB_PRESET_BD = Column(String(256))
    DEST_EXT = Column(String(10))
    HANDBRAKE_CLI = Column(String(256))
    MAINFEATURE = Column(String(10))
    HB_ARGS_DVD = Column(String(256))
    HB_ARGS_BD = Column(String(256))
    EMBY_REFRESH = Column(String(10))
    EMBY_SERVER = Column(String(256))
    EMBY_PORT = Column(String(10))
    NOTIFY_RIP = Column(String(10))
    NOTIFY_TRANSCODE = Column(String(10))
    MAX_CONCURRENT_TRANSCODES = Column(String(10))
    # Sensitive fields tracked for filtering
    EMBY_API_KEY = Column(String(256))
    IFTTT_KEY = Column(String(256))
    PB_KEY = Column(String(256))
    OMDB_API_KEY = Column(String(256))
    TMDB_API_KEY = Column(String(256))
    PO_USER_KEY = Column(String(256))
    PO_APP_KEY = Column(String(256))
    APPRISE = Column(String(1024))

    job = relationship("Job", back_populates="config")


# Sensitive fields to mask in API responses
HIDDEN_CONFIG_FIELDS = {
    "EMBY_API_KEY",
    "IFTTT_KEY",
    "PB_KEY",
    "OMDB_API_KEY",
    "TMDB_API_KEY",
    "PO_USER_KEY",
    "PO_APP_KEY",
    "APPRISE",
}


class AppState(Base):
    __tablename__ = "app_state"

    id = Column(Integer, primary_key=True)
    ripping_paused = Column(Boolean, default=False)


class SystemDrives(Base):
    __tablename__ = "system_drives"

    drive_id = Column(Integer, primary_key=True)
    name = Column(String(100))
    mount = Column(String(100))
    job_id_current = Column(Integer)
    job_id_previous = Column(Integer)
    description = Column(String(200))
    drive_mode = Column(String(100))
    maker = Column(String(25))
    model = Column(String(50))
    serial = Column(String(25))
    connection = Column(String(5))
    read_cd = Column(Boolean)
    read_dvd = Column(Boolean)
    read_bd = Column(Boolean)
    firmware = Column(String(10))
    location = Column(String(255))
    stale = Column(Boolean)
    mdisc = Column(Integer)
    serial_id = Column(String(100))


class SystemInfo(Base):
    __tablename__ = "system_info"

    id = Column(Integer, primary_key=True)
    name = Column(String(100))
    cpu = Column(String(20))
    description = Column(String(256))
    mem_total = Column(Float)


class Notifications(Base):
    __tablename__ = "notifications"

    id = Column(Integer, primary_key=True)
    title = Column(String(256))
    message = Column(Text)
    trigger_time = Column(DateTime)
    seen = Column(Boolean, default=False)
    cleared = Column(Boolean, default=False)
